### PR TITLE
Create Feature Feedback Widget

### PR DESCRIPTION
**Pull request by jbuckner**

### CODE REVIEW

File: src/feature-feedback.ts [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+{
+  @property({ type: String }) featureIdentifier?: string;
+
+  @property({ type: String }) prompt = 'Do you find this feature useful?';
+
+  @property({ type: String }) buttonText = 'Beta';
+
+  @property({ type: Object }) recaptchaManager?: RecaptchaManagerInterface;
+
+  @property({ type: Object }) resizeObserver?: SharedResizeObserverInterface;
```
Comment (iisa) [inline src/feature-feedback.ts:39]: ts q: what is the difference between setting this type w/ an interface instead of using `instance SharedResizeObserver`

File: src/feature-feedback.ts [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+              <div id="prompt-text">${this.prompt}</div>
+              <button
+                @click=${(e: Event) => {
+                  e.preventDefault();
+                  this.vote = this.vote === 'up' ? undefined : 'up';
+                }}
+                ?disabled=${this.processing}
+                class="vote-button upvote-button ${this
+                  .upvoteButtonClass} ${this.voteNeedsChoosing ? 'error' : ''}"
+                tabindex="0"
```
Comment (iisa) [inline src/feature-feedback.ts:245]: TIL: safari won't focus buttons even with an explicitly set `tab index` attribute
image

and here I thought that buttons have a default of `tabindex=0` 😄

File: src/feature-feedback.ts [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+              >
+                ${thumbsUp}
+              </button>
+              <button
```
Comment (iisa) [inline src/feature-feedback.ts]: q: was there a particular reason why we didn't use radio buttons?

**Review by iisa — APPROVED** [review]
looks good! just peppered some qs, comments while reviewing ^_^

also, just noting that the test file is empty.

File: src/feature-feedback.ts [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+
+  private disconnectResizeObserver(observer?: SharedResizeObserverInterface) {
+    observer?.removeObserver({
+      handler: this,
+      target: this.resizingElement,
+    });
+  }
+
+  private setupScrollObserver() {
+    document.addEventListener('scroll', this.boundScrollListener!);
```
Comment (iisa) [inline src/feature-feedback.ts]: q: what does the `!` at the end do?

File: src/feature-feedback.ts (hunk 1 above)
Comment (jbuckner) [reply src/feature-feedback.ts:39]: This is where the power of dependency injection comes into play! By using an interface instead of the actual object type, we can substitute the "real" `SharedResizeObserver` with a different implementation of it, ie say someone writes a more efficient version, or a mock `SharedResizeObserver` that specifically acts like we want it to during testing. ie:
```ts
// test.ts
export class MockSharedResizeObserver implements SharedResizeObserverInterface {
  // implement the SharedResizeObserverInterface to do whatever you need
}

const mockRO = new MockSharedResizeObserver();
<feature-feedback .resizeObserver=${mockRO}></feature-feedback>
```
The `<feature-feedback>` doesn't care what class it's actually given, as long as it conforms to the `SharedResizeObserverInterface`

File: src/feature-feedback.ts (hunk 4 above)
Comment (jbuckner) [reply src/feature-feedback.ts]: `this.boundScrollListener` is marked optional, but will effectively always have a value because we set it during `firstUpdated` so it's never really going to be undefined in practice. Using the `!` is just telling Typescript that that property will always be defined do you don't have to do `if (this.boundScrollListener) ...`.

Having said that, I think I can move the `!` up to the `boundScrollListener` definition so I don't need to do it here.

File: src/feature-feedback.ts (hunk 3 above)
Comment (jbuckner) [reply src/feature-feedback.ts]: No, but that's a good idea since they're mutually exclusive
